pkg/embeddedcluster/upgrade.go

```diff
@@ -0,0 +1,347 @@
+	}
+
+	log.Printf("Pulling upgrade binary from %s...", repo)
+
+	err = pullFromRegistry(ctx, k8sClient, repo, tmpdir)
+	if err != nil {
+		return "", fmt.Errorf("pull from registry: %w", err)
+	}
+
+	return filepath.Join(tmpdir, upgradeBinaryOperatorBinary), nil
+}
+
+const (
+	localArtifactMirrorMetadataKey = "local-artifact-mirror-image"
+)
+
+func getLocalArtifactMirrorImage(ctx context.Context, k8sClient kubernetes.Interface, in *embeddedclusterv1beta1.Installation) (string, error) {
+	path, err := pullEmbeddedClusterMetadataFromRegistry(ctx, k8sClient, in.Spec.Artifacts.EmbeddedClusterMetadata)
+	if err != nil {
+		return "", fmt.Errorf("pull embedded cluster metadata from registry: %w", err)
+	}
+
+	f, err := os.Open(path)
+	if err != nil {
+		return "", fmt.Errorf("open metadata file: %w", err)
+	}
+	defer f.Close()
+
+	var metadata embeddedclustertypes.ReleaseMetadata
+	err = json.NewDecoder(f).Decode(&metadata)
+	if err != nil {
+		return "", fmt.Errorf("decode metadata: %w", err)
+	}
+
+	srcImage, ok := metadata.Artifacts[localArtifactMirrorMetadataKey]
+	if !ok {
+		return "", fmt.Errorf("missing local artifact mirror image in embedded cluster metadata")
+	}
+
+	domain, _, _ := kotsutil.GetEmbeddedRegistryCreds(k8sClient)
+
+	imageName, err := embeddedRegistryImageAddress(domain, srcImage)
+	if err != nil {
+		return "", fmt.Errorf("get image address: %w", err)
+	}
+
+	return imageName, nil
+}
+
+func pullEmbeddedClusterMetadataFromRegistry(ctx context.Context, k8sClient kubernetes.Interface, repo string) (string, error) {
+	tmpdir, err := os.MkdirTemp("", "embedded-cluster-artifact-*")
+	if err != nil {
+		return "", fmt.Errorf("create temp dir: %w", err)
+	}
+
+	log.Printf("Pulling version metadata from %s...", repo)
+
+	err = pullFromRegistry(ctx, k8sClient, repo, tmpdir)
+	if err != nil {
+		return "", fmt.Errorf("pull from registry: %w", err)
+	}
+
+	return filepath.Join(tmpdir, "version-metadata.yaml"), nil
+}
+
+func pullFromRegistry(ctx context.Context, k8sClient kubernetes.Interface, srcRepo string, dstDir string) error {
+	domain, username, password := kotsutil.GetEmbeddedRegistryCreds(k8sClient)
+	store := credentials.NewMemoryStore()
+	store.Put(ctx, domain, auth.Credential{
+		Username: username,
+		Password: password,
+	})
+
+	transp, ok := http.DefaultTransport.(*http.Transport)
+	if !ok {
+		return fmt.Errorf("default transport is not http.Transport")
+	}
+	transp = transp.Clone()
+	transp.TLSClientConfig = &tls.Config{InsecureSkipVerify: true}
+
+	opts := pullArtifactOptions{}
+	opts.client = &auth.Client{
+		Client:     &http.Client{Transport: transp},
+		Credential: store.Get,
+	}
+
+	err := pullArtifact(ctx, srcRepo, dstDir, opts)
+	if err != nil {
+		return fmt.Errorf("pull oci artifact: %w", err)
+	}
+
+	return nil
+}
+
+func embeddedRegistryImageAddress(registryDomain string, srcImage string) (string, error) {
+	imagePath, err := image.PathToRegistryECImage(srcImage, registryDomain)
```

Comment:
the second parameter is the registry namespace, not the domain. these registry settings should be passed by the deploy function from kots.